### DISCUSSION TITLE
How to handle dynamically created pages in routing?

### DISCUSSION THREAD

**Discussion by OrangeTimes** [vikejs/vike#1661] [Discussion]
I'm creating +Page.js dynamically by fetching them from npm server using vike and Fastify server. When executing a request for the first time it fails with 404. The next request renders the page correctly as expected.  As I understand the internal router map is unaware of the newly created page on the filesystem. Here is the log when trying to render /test/ui-demo-page@5.9.9-test27

4:14:41 PM [vike][request(2)] HTTP request: /test/ui-demo-page@5.9.9-test27
4:14:41 PM [vike][config] created /test/ui-demo-page@5.9.9-test27/+Page.js
4:14:41 PM [vike][request(2)] URL /test/ui-demo-page@5.9.9-test27 doesn't match the route of any of your pages:

│ /                                     │ Filesystem Route │ /pages/                                │
│ /about                                │ Filesystem Route │ /pages/about/                          │
│ /star-wars                            │ Filesystem Route │ /pages/star-wars/index/                │
│ /star-wars/@id                        │ Filesystem Route │ /pages/star-wars/@id/                  │
│ /test                                 │ Filesystem Route │ /test/                                 │
│ /test/ui-demo-page@5.9.9-test26 │ Filesystem Route │ /test/ui-demo-page@5.9.9-test26/ │

See https://vike.dev/routing for more information about routing.
4:14:41 PM [vike][request(2)] HTTP response /test/ui-demo-page@5.9.9-test27 404

I have read the documentation on routing in vike but cannot figure out hot to handle this situation

**Comment by brillout** [comment]
Hm, I wouldn't know what the issue could be. The best would be a minimal reproduction.

**Reply by OrangeTimes** [reply]
[router.log](https://github.com/vikejs/vike/files/15486213/router.log)

Here I'm trying to debug.log the PageContext that I receive in router function 

/test/ui-demo-page@5.9.9-test29 is the original URL the file is present on filesystem test\re-ja-ui-demo-page@5.9.9-test29\+Page.js

As you can see from the log there is no pageId for this new version of the page only previous versions test28 and older are present. I'm creating a symlink on the fly during request but before rendering starts. Maybe you can suggest something

**Reply by OrangeTimes** [reply]
It seems that i need to update internal object manually pageContext._pageRoutes or refresh it somehow. Because during request time it seems there is no pageId for newly created page. 
https://github.com/vikejs/vike/blob/main/vike/shared/route/index.ts

If you have no ideas I can spend some time on reproduction

**Reply by brillout** [reply]
I don't know, it's difficult to say from what you share. AFAICT the best here would be a reproduction that is minimal.

**Reply by OrangeTimes** [reply]
[vike-ssr-reproduction.zip](https://github.com/vikejs/vike/files/15487580/vike-ssr-reproduction.zip)

You should be able to reproduce this
1. npm install
2. npm run dev
3. access any  unique page like  http://localhost:3000/test123 (1st time 404,  next time the counter should be rendered of page refresh)
Somehow if it's the initial  (cold) start of the server it doesn't work the exact way (I have no idea why) it only sends 404 always. You may need to restart it to be able to reproduce it (by trying another unique URL)

**Reply by brillout** [reply]
Is the reproduction minimal? Can you publish it on GitHub?

**Reply by OrangeTimes** [reply]
https://github.com/OrangeTimes/vike-ssr-dynamic-pages

Okay i have tried to simplify it to a minimal setup

**Reply by brillout** [reply]
Isn't it a race condition? You're generating pages upon HTTP requests; it may take a while until Vite/Vike picks up the new page (HMR isn't instantaneous). Isn't it expected that you need to wait a little between the time you generate the page and the time you can access the page?

What's your use case for trying to generate pages on the fly? :eyes:

**Reply by OrangeTimes** [reply]
Well, it' doesn't look like a classical race condition to me. How can I force Vike to pick up a new page? I have file path to it, I know it is already present on the server in the right directory with proper content so I know it is ready to be rendered. But vike doesn't have any knowledge of it and it seems there is no way do update vike's "routing table" with new information on the fly :-(

The use case is that different teams publish pages in a single NPM registry demo-page@0.0.1. We need a solution that will support the rendering of a newly published page on the fly without republishing the rendering solution, and redeploying it to production multiple times.  We have many different pages and their versions published during the day (the pages are developed and published in isolation). 
Another approach I was trying is to read the published page as a string,  pushing it to page context as a custom property but there are multiple problems while rendering this string on the client, and this approach doesn't look very clean to me...

**Reply by brillout** [reply]
I think the bottom line here is this:

>  it may take a while until Vite/Vike picks up the new page (HMR isn't instantaneous)

That's actually an inherent limitation: watching the filesystem isn't instantaneous. In other words: I don't think any framework out there supports what you want.

So I'd recommend to go with another approach.

It's an interesting use case. If there is an interest on your side, then there may be an opportunity to have some kind of [sponsorship-based](https://github.com/vikejs/vike/issues/1350) collaboration.

**Reply by OrangeTimes** [reply]
I don't want to make it happen automatically, I just need a way to inject a new file route into Vike programmatically during request time. or maybe using route function

Regarding sponsorship, it's hard to commit to anything because we are not sure how Vike fits our needs :-(  we are just at PoC state. So far it was just too hard to make it work. What approach would you suggest? How to fetch a page and render it on the go?

**Reply by brillout** [reply]
> What approach would you suggest? How to fetch a page and render it on the go?

It's difficult to say without knowing more about your use case.

I'd suggest to keep trying (alternative approaches), while feel free to create new feature requests where you see fit.

**Reply by OrangeTimes** [reply]
Is it possible to add a new file route using the route function?

**Reply by brillout** [reply]
> Is it possible to add a new file route using the route function?

I think the docs are pretty clear that this isn't the purpose of Route Functions? See https://vike.dev/route-function.

Maybe a more fruitful discussion would be that you described the fundamental requirement you're trying to fulfill. Is it about scaling your teams? Are you a media agency and it's about some kind of customer requirement?

Your approach is quite unusual, and it's difficult to justify spending time adapting Vike for the peculiar setup of a single user.

**Reply by OrangeTimes** [reply]
Ok we have a few teams that work on different versions of the same page for A/B experiments and we have many of these experiments running in production simultaneously. The page can have multiple variations, all of which are published to NPM registry as separate isolated packages  It allows teams to work independently and also allows us to scale them as well. We have a separate router service that determines in runtime which page the customer should go to based on criteria set in router config and customer request headers like GeoIP and others. Later, the router calls our existing SSR solution to render specific pages and versions and return the pre-rendered content to the customer. We want to modernize our SSR solution (developed more than 8 years ago) and thought that something on the market can already do some stuff we do and do it better and provide more features to us (latest React support, prefetching etc). So far the teams are really happy working in such a setup so we don't want to change things for them. Also, we need a single solution and not like 5 different apps for SSR because we are limited on resources (each team can request multiple copies of the production environment) If we add a lot of apps for SSR we will run out of resources pretty quickly. Also publishing pages in the scope of the SSR app will not work for us because we don't want to redeploy the app multiple times on prod affecting caching. After all, teams can publish a new page or variation of it every hour. So it we need a solution that can be deployed as single instance that can render all our pages on the fly without the need of redeployment to support some new pages. 

We have a tested way how to fetch a page on the filesystem on the go using pnpm store, so I was thinking in 2  approaches
 - create symlink like /pages/demopage@1.0.0/Page+ to pnpm store on the go during the initial request to VIKE. But it seems there is no way to tell Vike about new file route programmatically 
 - parse bundled page as a string and pass it to PageContext as some custom property. It was working for server rendering, but I didn't manage to make it work during hydration due to different errors on the client

**Reply by brillout** [reply]
Quickly glazing over your answer, it seems like your only blocker is your hack of mis-using npm packages to re-deploy. AFAICT all your other concerns are usual use cases and can be achieved with Vike and without any quirks.

I'd be happy to help about your other concerns (a lot of what you wrote seem unrelated to the problem of the original post) but I think the best would be some kind of sponsoring or consulting.

Precise, well-written, and well-explained feature request(s) welcome. (Regardless of sponsoring.)
